Dump departmental data as CSV files

tomx4096 opened the pull request:
## Status

Ready for review / in progress

## Description of Changes

Fixes #489 

Changes proposed in this pull request:

 - Download a department's officer data as CSV
 - Download a department's incident data (if available) as CSV
 - Unit tests

## Notes for Deployment

The final presentation and formatting of data in the CSVs is undecided and open to feedback. Better __str__ methods in some of the model objects would improve presentation

## Screenshots (if appropriate)
![screenshot from 2018-08-08 07-21-10](https://user-images.githubusercontent.com/18009772/43817843-dd6164e2-9adb-11e8-95f1-7dd8bf9f9588.png)
![screenshot from 2018-08-08 07-21-44](https://user-images.githubusercontent.com/18009772/43817844-dd82b476-9adb-11e8-82b8-40ef49a60415.png)



## Tests and linting

 - [x] I have rebased my changes on current `develop`

 - [x] pytests pass in the development environment on my local machine

 - [x] `flake8` checks pass

redshiftzero commented on OpenOversight/app/main/views.py:
```diff
@@ -565,6 +565,85 @@ def submit_department_images(department_id=1):
+            assign_dict[r.officer_id] = []
+        assign_dict[r.officer_id].append("(#%s %s %s %s %s)" % (check_input(r.star_no), check_input(r.rank), check_input(r.unit), check_input(r.star_date), check_input(r.resign_date)))
+
+    notes_dict = {}
```
Let's remove the notes from the CSV, the reason being is that the notes are actually for admins/area coordinators to add private notes to a profile, so we don't want to expose them to regular users

redshiftzero commented on OpenOversight/app/main/views.py:
```diff
@@ -565,6 +565,85 @@ def submit_department_images(department_id=1):
+                    check_input(record.report_number),
+                    check_input(record.date),
+                    check_input(record.description),
+                    check_input(record.address),
```
Let's add a string representation for `Location` so the output here is nice, i.e.:

```
diff --git a/OpenOversight/app/models.py b/OpenOversight/app/models.py
index c4b944b..9e87a13 100644
--- a/OpenOversight/app/models.py
+++ b/OpenOversight/app/models.py
@@ -222,6 +222,20 @@ class Location(db.Model):
     def validate_state(self, key, state):
         return state_validator(state)
 
+    def __repr__(self):
+        if self.street_name and self.cross_street2:
+            return 'Intersection of {} and {}, {} {}'.format(
+                self.street_name, self.cross_street2, self.city, self.state)
+        elif self.street_name and self.cross_street1:
+            return 'Intersection of {} and {}, {} {}'.format(
+                self.street_name, self.cross_street1, self.city, self.state)
+        elif self.street_name and self.cross_street1 and self.cross_street2:
+            return 'Intersection of {} between {} and {}, {} {}'.format(
+                self.street_name, self.cross_street1, self.cross_street2,
+                self.city, self.state)
+        else:
+            return '{} {}'.format(self.city, self.state)
+
```

redshiftzero commented on OpenOversight/app/templates/browse.html:
```diff
@@ -14,14 +14,20 @@ <h1><small>Browse a Department</small></h1>
+          <a class="btn btn-lg btn-primary" href="{{ url_for('main.download_dept_csv', department_id=department.id) }}">
+               Download Officer Data
+          </a>
+          </p>
```
Not necessary for merge on this PR, but just a thought: we should have the officers and incidents links be primary, and the raw data downloads be somewhere less central. One good place on the site might be a link on the footer to a new route that has the CSV download links for each department:

<img width="739" alt="screen shot 2018-08-18 at 6 52 23 pm" src="https://user-images.githubusercontent.com/7832803/44304690-dfda4880-a317-11e8-90e8-6284a9ca53ac.png">

If you feel like implementing this (or if you have another better idea!) feel free, else we can defer to a followup.

tomx4096 commented:
Thanks for the reviews @redshiftzero . I removed the notes from the CSV, added your Location representation, and changed the location of the download links to the footer:
![screenshot from 2018-08-20 11-27-57](https://user-images.githubusercontent.com/18009772/44332850-66ed0500-a46d-11e8-977d-865a7e0d3767.png)
![screenshot from 2018-08-20 11-27-48](https://user-images.githubusercontent.com/18009772/44332853-6b192280-a46d-11e8-8956-e1b5b63f6b8a.png)
